lib/luaT/luaT.c

```diff
@@ -471,53 +471,136 @@ void luaT_registeratname(lua_State *L, const struct luaL_Reg *methods, const cha
 }
 
 
-/* utility functions */
-const char *luaT_classrootname(const char *tname)
+/* returns the name of the class itself (sans nesting) */
+const char* luaT_classrootname(const char *tname)
 {
-  int i;
+  int idx;
   int sz = strlen(tname);
 
-  for(i = 0; i < sz; i++)
+  for(idx = sz-1; idx >= 0 ; idx--)
   {
-    if(tname[i] == '.')
-      return tname+i+1;
+    if(tname[idx] == '.')
+      return tname+idx+1;
   }
   return tname;
 }
 
-/* module_name must be a buffer at least as big as tname 
- * return true if the class is part of a module */
-int luaT_classmodulename(const char *tname, char *module_name)
+/* parent_name must be a buffer at least as big as tname.
+ * If class has a parent, returns true; and, sets
+ * parent name to that of full parent hierarchy (e.g.
+ * given class `A.b.c`, sets parent_name to `A.b`)
+ */
+int luaT_fullparentname(const char *tname, char *parent_name)
+{
+  int sz = strlen(tname);
+  int idx;
+  for(idx = sz-1; idx >= 0 ; idx--)
+    if(tname[idx] == '.') break;
+
+  if (idx == 0) return 0;
+
+  strncpy(parent_name, tname, idx);
+  parent_name[idx] = '\0';
+  return 1;
+}
+
+/* parent_name must be a buffer at least as big as tname.
+ * If class has a parent, returns true; and, sets
+ * parent name to that of outermost parent (e.g.
+ * given class `A.b.c`, sets parent_name to `A`)
+ */
+int luaT_outerparentname(const char *tname, char *parent_name)
 {
   char chars[] = {'.', '\0'};
   size_t n;
   n = strcspn(tname, chars);
-  strncpy(module_name, tname, n);
-  module_name[n] = '\0';
+  strncpy(parent_name, tname, n);
+  parent_name[n] = '\0';
   return tname[n] == '.';
 }
 
-/* Lua only functions */
+/* parent_name must be a buffer at least as big as tname.
+ * If class has a parent, returns true; and, sets
+ * parent name to that of innermost parent (e.g.
+ * given class `A.b.c`, sets parent_name to `b`)
+ */
+
+int luaT_innerparentname(const char *tname, char *parent_name)
+{
+  int sz = strlen(tname);
+  int tail, head;
+  for(tail = sz-1; tail >= 0 ; tail--)
+    if(tname[tail] == '.') break;
+
+  if (tail == 0) return 0;
+
+  for(head = tail; head >= 1; head--)
+    if(tname[head-1] == '.') break;
+
+  strncpy(parent_name, tname+head, tail-head);
+  parent_name[tail-head] = '\0';
+  return 1;
+}
+
+void traverse_tables(lua_State *L, const char *tname)
```

Comment:
Can you give this a more descriptive name? Maybe something like `loadinnerparent`?